File: algorithms/graph/maximum-bipartite-matching/c/maximum_bipartite_matching.c
#include "maximum_bipartite_matching.h"
#include <string.h>

#define MAX_V 500
static int adj[MAX_V][MAX_V], adj_count[MAX_V];
static int match_right[MAX_V], visited[MAX_V];

static int dfs(int u) {
    for (int i = 0; i < adj_count[u]; i++) {
        int v = adj[u][i];
        if (!visited[v]) {
            visited[v] = 1;
            if (match_right[v] == -1 || dfs(match_right[v])) {
                match_right[v] = u;
                return 1;
            }
        }
    }
    return 0;
}

int maximum_bipartite_matching(int arr[], int size) {
    int nLeft = arr[0], nRight = arr[1], m = arr[2];
    memset(adj_count, 0, sizeof(int) * nLeft);
    memset(match_right, -1, sizeof(int) * nRight);
    for (int i = 0; i < m; i++) {
        int u = arr[3 + 2 * i], v = arr[3 + 2 * i + 1];
        adj[u][adj_count[u]++] = v;
    }
    int result = 0;
    for (int u = 0; u < nLeft; u++) {
        memset(visited, 0, sizeof(int) * nRight);
        if (dfs(u)) result++;
    }
    return result;
}

File: algorithms/graph/maximum-bipartite-matching/rust/maximum_bipartite_matching.rs
pub fn maximum_bipartite_matching(arr: &[i32]) -> i32 {
    let n_left = arr[0] as usize;
    let n_right = arr[1] as usize;
    let m = arr[2] as usize;
    let mut adj = vec![vec![]; n_left];
    for i in 0..m {
        let u = arr[3 + 2 * i] as usize;
        let v = arr[3 + 2 * i + 1] as usize;
        adj[u].push(v);
    }
    let mut match_right = vec![-1i32; n_right];

    fn dfs(u: usize, adj: &[Vec<usize>], match_right: &mut [i32], visited: &mut [bool]) -> bool {
        for &v in &adj[u] {
            if !visited[v] {
                visited[v] = true;
                if match_right[v] == -1 || dfs(match_right[v] as usize, adj, match_right, visited) {
                    match_right[v] = u as i32;
                    return true;
                }
            }
        }
        false
    }

    let mut result = 0i32;
    for u in 0..n_left {
        let mut visited = vec![false; n_right];
        if dfs(u, &adj, &mut match_right, &mut visited) { result += 1; }
    }
    result
}
